src/algebra/algebra/tower.lean

```diff
@@ -122,8 +122,30 @@ ring_hom.ext $ λ _, rfl
 @[simp] lemma coe_to_alg_hom' : (to_alg_hom R S A : S → A) = algebra_map S A :=
 rfl
 
+section biproduct
+
+variables {T : Type*} [comm_semiring S] [comm_semiring A] [comm_semiring B] [comm_semiring T]
+variables [algebra S A] [algebra S B] [algebra S T] [algebra A T] [algebra B T]
+variables {A B}
+variables (S T)
+
+
+-- TODO: move this
+def algebra.biprod [is_scalar_tower S A T] [is_scalar_tower S B T] :
+  A →ₗ[S] B →ₗ[S] T := sorry
+-- ((algebra.lmul S T).to_linear_map.compl₂ ((algebra.linear_map B T).restrict_scalars S)).comp
+--   ((algebra.linear_map A T).restrict_scalars S)
```

Comment:
`restrict_scalars` isn't defined till line 207

Reply:
Thanks! It was as stupid as that. I have moved it, and it seems to work. If you have any suggestion on some doc to be added there (insted of the temporary Move This), I can implement it, otherwise I'd simply leave it as it is.

Reply:
Frankly I think this whole biprod thing is a distraction; unless you can come up with a motivation for it, I'd be inclined to spin off a new PR and forget about it until someone indicates they want it.

If you replace the uses of `algebra.biprod` with `algebra.lmul`, your proofs become much simpler and yet also more general.

Reply:
To be honest, I did not have any `algebra.biprod` in my first version of the file, it has recently appeared because you proposed this `def algebra.biprod`, which I accepted to follow. I only had the `kronecker_biprod` following a suggestion made by Adam Topaz on Zulip, that the general version of two algebras mapping to the third would have been the easiest to work with if one had in mind the tensor product. It was meant to be a step helping the proof of the associativity of the "usual" Kronecker product (the only one for which I suggested a symbol). You proposed a lot of useful and very beneficial suggestions, but I don't think that this Kronecker product would deserve so much time spent on it and I am not going to spend much more time on it. I am perfectly happy to have this PR archived waiting for someone else willing to start with a different approach, hopefully taking advantage of what has been done here.

Reply:
Sure, thanks for all your work. I've split off a chunk of this PR at #8560 - would you mind taking a look?

Reply:
How much of this PR remains after #8560? Maybe I should just wait till that one hits master.

Reply:
What remains is mainly associativity, I guess, even now that #8560 is in master.

Reply:
I've pushed a merge commit that should remove the duplicated parts. I haven't attempted to move things to new homes.

Reply:
Ok, I am happy to have a look at this file tomorrow morning (am a bit busy right now). At any rate, what do you suggest to do? My instinct would be to extract my "old" proof of associativity and add it to the file you created in #8560 (getting rid of `matrix/kronecker_product_bialgebra`), what do you think? Or do you have "new homes" in mind which you would like me to build?

Reply:
Unless you see a use for `algebra.biprod`, my feeling would be to do as you suggest, but in a new PR. This one can be abandoned until someone decides they need the biprod versions.

Reply:
OK!

Reply:
Ok, I have PRed in #8700 two lemmas about associativity. Concerning this `PR`, can I delete this git branch or would this destroy the PR?

Reply:
I think it will destroy the PR, but not irreversibly.

Reply:
Ok, then I leave the branch. I was just trying to do some house-keeping.

Reply:
@faenuccio What is your plan with the rest of this PR? Was #8560 enough for LTE purposes? Should we refactor LTE now?

Reply:
@jcommelin #8560 was not enough, because we lacked associativity. I have opened #8700 to add this and @eric-wieser is reviewing it. I believe that #8700 will be enough for LTE purposes, so my plan is, once it will land in `master`, to `bump mathlib` to `LTE` and the refactor it accordingly.